### ISSUE TITLE
Optimize compiling time for MCMC

### ISSUE THREAD

**Issue by fehiepsi** [pyro-ppl/numpyro#585]
Now we start looking at large models. We need to remove unnecessary traveling the model to reduce compiling time.

#### Single-chain init

We travel 4 times: 1 to find `inv_transforms`, 1 to find init sampling state (which involves computing potential energy), 2 to find valid initial params (which involves computing both `inv_transforms` and potential energy).

- [x] We should only need to travel 2 times here: we can use the output of `find valid initial params` to define `inv_transform` and set the init sampling state (because `find_valid_initial_params` also computes pe and its grad)

#### Single-chain sample

- [x] After #584 and a small change to `get_potential_fn`, we travel 1 time for each sampling step to compute potential energy. This is expected!

#### Multi chain

- [ ] Currently, we do `pmap(init + sample)`. But it is better to do `pmap/map/vmap(init) + pmap(sample)` because `init + sample` produces much larger jaxpr than each single one and JAX compiling time is nonlinear w.r.t. jaxpr length.

#### heuristic step size

- [x] (pretty low priority) probably we can also find heuristic step size with only 1 time for compiling (if users want to use this functionality)

**Comment by fehiepsi** [comment]
Currently, using `pmap(init + sample)` really simplifies a lot of code for us. So probably the last task is unnecessary unless we face some slow compiling issues again.

**Comment by mcguenther** [comment]
> `pmap`

o pmap
